pageserver/src/tenant/remote_timeline_client/upload.rs

```diff
@@ -56,9 +58,22 @@ pub(super) async fn upload_timeline_layer<'a>(
     });
     let storage_path = conf.remote_path(source_path)?;
 
-    let source_file = fs::File::open(&source_path)
-        .await
-        .with_context(|| format!("Failed to open a source file for layer {source_path:?}"))?;
+    let source_file_res = fs::File::open(&source_path).await;
+    let source_file = match source_file_res {
+        Ok(source_file) => source_file,
+        Err(e) if e.kind() == ErrorKind::NotFound => {
+            // In some situations we might run into the underlying file being deleted by
+            // e.g. compaction before the uploader gets to it. In that instance, we don't
+            // want to retry the error: a deleted file won't come back. In theory, the
+            // file might not have been written in the first place, which also indicates
+            // a bug. Still log the situation so that we can keep an eye on it.
+            // See https://github.com/neondatabase/neon/issues/4526
```

Comment:
In general we are quite good on writing stuff to disk *before* scheduling an upload.

However ... wondering do we need this much text because we also log. From the point of view of remote_timeline_client, this is very much a persistent failure and we can't do anything about it.

On the actual site of the race, we could add some "FIXME: between `create_delta_layer` and scheduling an upload, a compaction could come in and delete the file". Mentioning that, I was thinking we would have sort of preconditions on scheduling a file upload, that you couldn't schedule a not found path, but I guess we don't have such? Confirmed we don't. Unsure if that is actionable.

Reply:
I think that if this is not actionable, we can either remove the log or move it to trace level

Reply:
> we can either remove the log or move it to trace level

Aah agreed yeah ... Yeah, perhaps we should make it into a `debug`, as we cannot do anything. We still see the layer being flushed, and layer being deleted by compaction in the logs, we just wouldn't see the (later) error at all about it not being uploaded.

Reply:
I have not looked it up on grafana but IIRC the race occurs only very rarely? It's just that before it blocked the upload queue so each time it happened it had a non trivial impact.

I think it is useful to have at least some message here. In the future due to some rearchitecting it might occur that we delete files that are actually still useful and then we might wonder why, even though they are scheduled for upload, are not uploaded.

I mean in theory the uploader might also be given wrong file names, and that's why they don't exist, or the files haven't been written to disk yet. Or there has been a restart and the filesystem with the layer files is not mounted yet, etc (in the future one might want to do that, I'm not saying it should be done, just that we are ideally future proof here).